### PR TITLE
Feature/lock configurable random times

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Trello card: https://trello.com/c/IJdLfTVw/5237-improve-mongodb-lock-retries
Improve locking mechanism with:
- random sleeps between lock acquire/release retries (uniform distribution between configurable min and max values)
- keep track of latest release times and counts, to
- define default configuraiton, allowing to override any required value

### How to review

- Make sure code changes make sense
- Make sure unit tests pass

(Info) tested with dataset API as follows:
  - Added debug logs in `dp-mongoDB` to see the Usages map (with instanceID and callers)
  - Use config values that can be noticeable and testable at human scale (seconds instead of milliseconds):
```
	// Create MongoDB lock client, which also starts the purger loop
	th := 3 * time.Second
	sl := 5 * time.Second
	m.lockClient = dpMongoLock.New(ctx, m.Session, m.Database, instanceCollection, &dpMongoLock.ConfigOverride{
		TimeThresholdSinceLastRelease: &th,
		UsageSleep:                    &sl,
	})
```
  - Pass the CallerID or User stored in context as unique caller name: 
```
  func (m *Mongo) AcquireInstanceLock(ctx context.Context, instanceID string) (lockID string, err error) {
	uniqueCallerName := request.Caller(ctx)
	if uniqueCallerName == "" {
		uniqueCallerName = request.User(ctx)
	}
	log.Info(ctx, "$$$$$$$$$$$$$$$", log.Data{"uniqueCallerName": uniqueCallerName})
	return m.lockClient.Acquire(ctx, instanceID, uniqueCallerName)
}
```
  - Tried a cantabular import (with dp-compose), and checked that it was successul - note that the caller is the owner of the service token in this case.
  - Tried to do a `PUT /instances/{id}` with a florence token - note that the caller is the florence email in this case.
  - Tried to do the same `PUT /instances/{id}` 10 times - see that postman was locked for 5 seconds
  - Tried the same `PUT /instances/{id}`: it was fine, and the counter was reset
  - Waited just over 5 minutes (purger period)
  - See that old entries were removed from the map, as expted

### Who can review

Anyone

### CODE REVIEW

File: dplock/usage.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	if !found {
+		return
+	}
+	if usage.Count >= MaxCount && time.Since(usage.Released) <= TimeThresholdSinceLastRelease {
+		Sleep(UsageSleep)
+		usage.Count = 0
+	}
+}
+
+var Sleep = func(d time.Duration) {
```
Comment (redhug1) [inline dplock/usage.go:108]: I don't think this needs to be a separate function and could be inlined as it is only called from line 60

**Review by redhug1 — COMMENTED** [review]
see single observation for now ...

File: dplock/usage.go (hunk 1 above)
Comment (DavidSubiros) [reply dplock/usage.go:108]: I had to create a wrapper function to be able to create unit tests without sleeping, and validating the sleep time (i.e. we don't want unit tests to sleep, and we want them to check that the code called sleep instead).

example, in usage_test.go we do:
```
		slept := []time.Duration{}
		dplock.Sleep = func(d time.Duration) {
			slept = append(slept, d)
		}
```

and then validate that the slept slice contains what we expect

**Review by redhug1 — APPROVED** [review]
Tests: OK
Changes: make sense

LGTM

**Pull request merged**